src/combinatorics/additive/salem_spencer.lean

```diff
@@ -0,0 +1,386 @@
+multiplicative average of any two distinct elements is not in the set. -/
+@[to_additive "A Salem-Spencer, aka non averaging, set `s` in an additive monoid
+is a set such that the average of any two distinct elements is not in the set."]
+def mul_salem_spencer : Prop := ∀ ⦃a b c⦄, a ∈ s → b ∈ s → c ∈ s → a * b = c * c → a = b
+
+/-- Whether a given finset is Salem-Spencer is decidable. -/
+@[to_additive]
+instance {α : Type*} [decidable_eq α] [monoid α] {s : finset α} :
+  decidable (mul_salem_spencer (s : set α)) :=
+decidable_of_iff (∀ a ∈ s, ∀ b ∈ s, ∀ c ∈ s, a * b = c * c → a = b)
+  ⟨λ h a b c ha hb hc, h a ha b hb c hc, λ h a ha b hb c hc, h ha hb hc⟩
+
+variables {s t}
+
+@[to_additive]
+lemma mul_salem_spencer.mono (h : t ⊆ s) (hs : mul_salem_spencer s) : mul_salem_spencer t :=
+λ a b c ha hb hc, hs (h ha) (h hb) (h hc)
+
+@[to_additive, simp]
+lemma mul_salem_spencer_empty : mul_salem_spencer (∅ : set α) := λ x _ _ hx, hx.elim
+
+@[to_additive]
+lemma mul_salem_spencer.prod {t : set β} (hs : mul_salem_spencer s) (ht : mul_salem_spencer t) :
+  mul_salem_spencer (s.prod t) :=
+λ a b c ha hb hc h,
+  prod.ext (hs ha.1 hb.1 hc.1 (prod.ext_iff.1 h).1) (ht ha.2 hb.2 hc.2 (prod.ext_iff.1 h).2)
+
+@[to_additive]
+lemma mul_salem_spencer_pi {ι : Type*} {α : ι → Type*} [Π i, monoid (α i)] {s : Π i, set (α i)}
+  (hs : ∀ i, mul_salem_spencer (s i)) :
+  mul_salem_spencer ((set.univ : set ι).pi s) :=
+λ a b c ha hb hc h, funext $ λ i, hs i (ha i trivial) (hb i trivial) (hc i trivial) $ congr_fun h i
+
+end monoid
+
+section cancel_comm_monoid
+variables [cancel_comm_monoid α] {s : set α} {a : α}
+
+@[to_additive]
+lemma mul_salem_spencer.mul_left (hs : mul_salem_spencer s) : mul_salem_spencer ((*) a '' s) :=
+begin
+  rintro _ _ _ ⟨b, hb, rfl⟩ ⟨c, hc, rfl⟩ ⟨d, hd, rfl⟩ h,
+  rw [mul_mul_mul_comm, mul_mul_mul_comm a d] at h,
+  rw hs hb hc hd (mul_left_cancel h),
+end
+
+@[to_additive]
+lemma mul_salem_spencer.mul_right (hs : mul_salem_spencer s) :
+  mul_salem_spencer ((λ x, x * a) '' s) :=
+begin
+  rintro _ _ _ ⟨b, hb, rfl⟩ ⟨c, hc, rfl⟩ ⟨d, hd, rfl⟩ h,
+  rw [mul_mul_mul_comm, mul_mul_mul_comm d] at h,
+  rw hs hb hc hd (mul_right_cancel h),
+end
+
+@[to_additive]
+lemma mul_salem_spencer_mul_left_iff : mul_salem_spencer ((*) a '' s) ↔ mul_salem_spencer s :=
+⟨λ hs b c d hb hc hd h, mul_left_cancel (hs (set.mem_image_of_mem _ hb) (set.mem_image_of_mem _ hc)
+  (set.mem_image_of_mem _ hd) $ by rw [mul_mul_mul_comm, h, mul_mul_mul_comm]),
+  mul_salem_spencer.mul_left⟩
+
+@[to_additive]
+lemma mul_salem_spencer_mul_right_iff :
+  mul_salem_spencer ((λ x, x * a) '' s) ↔ mul_salem_spencer s :=
+⟨λ hs b c d hb hc hd h, mul_right_cancel (hs (set.mem_image_of_mem _ hb) (set.mem_image_of_mem _ hc)
+  (set.mem_image_of_mem _ hd) $ by rw [mul_mul_mul_comm, h, mul_mul_mul_comm]),
+  mul_salem_spencer.mul_right⟩
+
+end cancel_comm_monoid
+
+section nat
+
+lemma add_salem_spencer_iff_eq_right {s : set ℕ} :
```

Comment:
This should generalise a bit - I'm not sure what the appropriate generalisation is though

Reply:
Yeah... But we're missing `nsmul_left_cancel`. I would like it to be true in a `canonically_ordered_add_comm_monoid`, but your counterexample in archive shows it's not. So maybe `canonically_ordered_comm_semiring`?

Reply:
Note that we could take that as a definition. It's strictly stronger in the cases we don't care about, and equivalent in `ℕ`, `ℤ`...
I personally find it less elegant.

Reply:
We could just have the assumption that `2 *` is injective on `s`, which would be good enough for this.

Reply:
Just added a todo.